### DISCUSSION TITLE
VS Code extension "Goto Definition" for Cap'n Proto schema

### DISCUSSION THREAD

**Discussion by trickstar0301** [capnproto/capnproto#2124] [Ideas]
I'm working on a VS Code extension for definition jumping, but I'm unable to get the declaration position(file, line, column) from parsedSchema.

I suggest adding a new member to Node to reserve the declaration position, like this:

``` cpp
class Compiler::Node {
private:
  // ... existing members ...
  kj::Maybe<Declaration::Reader> rawDeclaration;
  // ... rest of the class ...
};
```
Additionally, I suggest exposing a way for library users to retrieve this information.
This would make it easier to retrieve the field's declaration position, improving integration with tools like VS Code.   
What do you think?

**Comment by kentonv** [comment]
I've always wanted to do this!

I agree that it would be ideal if we can use the canonical parser implementation for this, but that it doesn't provide all the information needed today. There's some partial support for tracking code locations but the information doesn't make its way all the way through the compiler. I'd love to work on this but realistically I don't expect to have time anytime soon. If you'd like to take a shot at it, feel free.

**Comment by trickstar0301** [comment]
@kentonv 

Thank you for your reply.
I will first add byte positions to Node to reserve the positions of declaration.

```
struct Node {
  id @0 :Id;
...
  startByte @34 :UInt32;
  endByte @35 :UInt32;
}
```

Does this approach look correct to you?
Or should I add byte position to SourceInfo?  
-> I decided to define it in SourceInfo according to Pull Request comment.

**Comment by trickstar0301** [comment]
@kentonv 

I have made a trial Pull Request for the changes to retain the position of declaration statements. 
Could you please take a look?
https://github.com/capnproto/capnproto/pull/2126
As for obtaining the position of an Expression, I currently have no idea. It would be helpful if the information were available in the ``MemberInfo`` struct in ``node-translator.c++``

**Comment by JonathanPlasse** [comment]
@trickstar0301 Have you considered contributing to @norgor VS Code extension [vscode-capnp](https://github.com/norgor/vscode-capnp) which seems to be the most recent and feature rich?

**Comment by trickstar0301** [comment]
@JonathanPlasse   
I’m still considering whether to contribute to `vscode-capnp`. First, I’d like to implement it as a language server that can be used outside of VSCode as well.

**Reply by JonathanPlasse** [reply]
Making an LSP for Cap'n Proto would be awesome. Let me know if you want some help, I have some experience with linters.

**Reply by ngorke** [reply]
Yeah that would be awesome. I was also thinking of an LSP, but I just have not had the time to do any research on this. 
Best of luck, and feel free to contribute to vscode-capnp!

**Reply by trickstar0301** [reply]
@norgor 
Thank you so much!
After making LSP server, I will let you know.

**Comment by trickstar0301** [comment]
@JonathanPlasse 

Thank you. That’s really encouraging!

**Comment by trickstar0301** [comment]
@kentonv 
Thank you for merging the [pull request](https://github.com/capnproto/capnproto/pull/2126). 
I’ve been working on a Language Server implementation for Cap'n Proto, which you can find here: [capnp-ls repository](https://github.com/trickstar0301/capnp-ls/tree/main). It currently supports basic LSP features like definition jumping and symbol analysis.

The implementation relies heavily on `kj` for overall design. Its efficiency and simplicity made it an excellent fit for building this.

Would you be open to discussing whether this could become part of the official Cap'n Proto repository? I believe it could greatly enhance the developer experience for users working with Cap'n Proto schemas.

@norgor
Please take a look at [sample client implementation](https://github.com/trickstar0301/capnp-ls/blob/main/samples/client/src/extension.ts). It demonstrates how to integrate with an LSP server.

**Comment by trickstar0301** [comment]
Hi @kentonv,

I hope you're doing well. I wanted to gently follow up regarding my language server, which adds features like definition jumping and symbol analysis. 
I understand you have a busy schedule, but if you could spare a moment to discuss its potential integration into the official repository, I would greatly appreciate it.

**Reply by kentonv** [reply]
Hi @trickstar0301, very sorry for being so slow to respond.

Your work is really cool and in an ideal world it should be part of capnp, maybe even built into the `capnp` tool itself. But realistically, I have a responsibility to carefully review any code that we bring into the core repo, and I just don't have the bandwidth to do it right now (obviously, given I don't even have bandwidth to reply to comments in a timely fashion...). So it's probably best for you to keep it in your own repo, at least for the time being.

I would be happy to link to your project on https://capnproto.org/otherlang.html -- feel free to submit a PR adding it to https://github.com/capnproto/capnproto/blob/master/doc/otherlang.md !

**Reply by trickstar0301** [reply]
Hi @kentonv,

Thank you for your response! I completely understand your decision, and I truly appreciate the time and support you've given me despite your busy schedule. 

I've created the PR as you suggested: [#2244](https://github.com/capnproto/capnproto/pull/2244).

Thanks again for your consideration!
